File: 027/remove-element.go
package main

import (
	"fmt"
)

func removeElement(nums []int, val int) int {
	// i: 第一个值为val的位置
	// j: 最后一个值不为val的位置
	i, j := 0, len(nums)-1
	for {
		for i < len(nums) && nums[i] != val {
			i++
		}

		for j >= 0 && nums[j] == val {
			j--
		}

		if i >= j {
			break
		}

		nums[i], nums[j] = nums[j], nums[i]
	}

	return i
}

func main() {
	nums := []int{3, 2, 2, 3}
	val := 3
	fmt.Println(removeElement(nums, val))
}

File: 027/RemoveElement.kt
class RemoveElement {
    fun removeElement(nums: IntArray, `val`: Int): Int {
        var non_val_size = 0
        for (i in nums)
            if (i != `val`)
                nums[non_val_size++] = i
        
        return non_val_size
    }
}